src/modules/ekf2/EKF/zero_innovation_heading_update.cpp

```diff
@@ -0,0 +1,84 @@
+/****************************************************************************
+ *
+ *   Copyright (c) 2022 PX4 Development Team. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in
+ *    the documentation and/or other materials provided with the
+ *    distribution.
+ * 3. Neither the name PX4 nor the names of its contributors may be
+ *    used to endorse or promote products derived from this software
+ *    without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+
+/**
+ * @file zero_innovation_heading_update.cpp
+ * Control function for ekf heading update when at rest or no other heading source available
+ */
+
+#include "ekf.h"
+
+void Ekf::controlZeroInnovationHeadingUpdate()
+{
+	const bool yaw_aiding = _control_status.flags.mag_hdg || _control_status.flags.mag_3D
+				|| _control_status.flags.ev_yaw || _control_status.flags.gps_yaw;
+
+	if (!_control_status.flags.tilt_align) {
+		// fuse zero heading innovation during the leveling fine alignment step to keep the yaw variance low
+		float innovation = 0.f;
+		float obs_var = _control_status.flags.vehicle_at_rest ? 0.001f : 0.1f;
+		fuseYaw(innovation, obs_var);
+
+		_last_static_yaw = NAN;
+
+	} else if (_control_status.flags.vehicle_at_rest) {
```

Comment:
```suggestion
	} else if (_control_status.flags.vehicle_at_rest && !yaw_aiding) {
```
I don't think we should fuse a fixed yaw if we have another yaw fusion active because they'll start to fight against each other.

Reply:
Already handled below on line 61. This still needs to run to set `_last_static_yaw` when at rest.